extras/gizmo/axis-shapes.js

```diff
@@ -0,0 +1,407 @@
+        this.axis = options.axis ?? 'x';
+        this._position = options.position ?? new Vec3();
+        this._rotation = options.rotation ?? new Vec3();
+        this._scale = options.scale ?? new Vec3(1, 1, 1);
+
+        this._defaultColor = options.defaultColor ?? Color.BLACK;
+        this._hoverColor = options.hoverColor ?? Color.WHITE;
+    }
+
+    hover(state) {
+        const material = state ? this._hoverColor : this._defaultColor;
+        for (let i = 0; i < this.meshInstances.length; i++) {
+            this.meshInstances[i].material = material;
+        }
+    }
+}
+
+class AxisArrow extends AxisShape {
+    _gap = 0;
+
+    _lineThickness = 0.04;
+
+    _lineLength = 0.5;
+
+    _arrowThickness = 0.15;
+
+    _arrowLength = 0.2;
+
+    constructor(options = {}) {
+        super(options);
+
+        this._createArrow(options.layers ?? []);
+    }
+
+    set gap(value) {
+        this._gap = value ?? 0;
+        this._updateLine();
+        this._updateArrow();
+    }
+
+    get gap() {
+        return this._gap;
+    }
+
+    set lineThickness(value) {
+        this._lineThickness = value ?? 1;
+        this._updateLine();
+        this._updateArrow();
+    }
+
+    get lineThickness() {
+        return this._lineThickness;
+    }
+
+    set lineLength(value) {
+        this._lineLength = value ?? 1;
+        this._updateLine();
+        this._updateArrow();
+    }
+
+    get lineLength() {
+        return this._lineLength;
+    }
+
+    set arrowThickness(value) {
+        this._arrowThickness = value ?? 1;
+        this._updateArrow();
+    }
+
+    get arrowThickness() {
+        return this._arrowThickness;
+    }
+
+    set arrowLength(value) {
+        this._arrowLength = value ?? 1;
+        this._updateArrow();
+    }
+
+    get arrowLength() {
+        return this._arrowLength;
+    }
+
+    _createArrow(layers) {
+        this.entity = new Entity('axis_' + this.axis);
+        this.entity.setLocalPosition(this._position);
+        this.entity.setLocalEulerAngles(this._rotation);
+        this.entity.setLocalScale(this._scale);
+
+        this._line = new Entity('line_' + this.axis);
+        this._line.addComponent('render', {
+            type: 'cylinder',
+            layers: layers,
+            material: this._defaultColor,
+            castShadows: false
+        });
+        this._updateLine();
+        this.entity.addChild(this._line);
+        this.meshInstances.push(...this._line.render.meshInstances);
+
+        this._arrow = new Entity('arrow_' + this.axis);
+        this._arrow.addComponent('render', {
+            type: 'cone',
+            layers: layers,
+            material: this._defaultColor,
+            castShadows: false
+        });
+        this._updateArrow();
+        this.entity.addChild(this._arrow);
+        this.meshInstances.push(...this._arrow.render.meshInstances);
+    }
+
+    _updateLine() {
+        this._line.setLocalPosition(new Vec3(0, this._gap + this._lineLength * 0.5, 0));
```

Comment:
Just incase this get's called frequently, is it worth avoiding the `new Vec3()`'s if possible

Reply:
Ahh yes I must have missed that one thanks!